### PR TITLE
python-xmltodict: add a new package

### PR DESCRIPTION
**Pull request by BKPepe**
Maintainer: me (@BKPepe)
Compile tested: Turris MOX, cortexa53, OpenWrt master and Turris Omnia, mvebu, OpenWrt SNAPSHOT r9420-c17a68c
Run tested: Turris MOX, cortexa53, OpenWrt master and Turris Omnia, mvebu, OpenWrt SNAPSHOT r9420-c17a68c

Proof of run tested:
Screenshot from 2019-04-21 03-34-20

Description:

- add [xmltodict](https://github.com/martinblech/xmltodict)

### CODE REVIEW

**Comment by BKPepe** [comment]
@commodo: Could you take a look if it is OK? :-)

File: lang/python/python-xmltodict/Makefile [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  TITLE:=Work with XML like JSON
+  URL:=https://github.com/martinblech/xmltodict
+  DEPENDS:= \
+	+python3-light \
+	+python3-xml \
+	+python3-urllib
+  VARIANT:=python3
+endef
+
+define Package/python3-aiohttp/description
```
Comment (commodo) [inline lang/python/python-xmltodict/Makefile]: typo: python3-aiohttp -> python3-xmltodict

Comment (BKPepe) [reply lang/python/python-xmltodict/Makefile]: I will change it, thank you.

**Comment by commodo** [comment]
rest looks good;

one question before merging; 
should we resolve a consensus about naming ?

i.e. whether it's `python3-xmltodict` or `python-xmltodict` all the way [except for places where it cannot be avoided like python3-package.mk, Py3Package, VARIANT:=python3, etc] ?

related to this comment: https://github.com/openwrt/packages/pull/8741#issuecomment-485657698

**Comment by BKPepe** [comment]
I think it would be better to stay with python-xmltodict as advised. I will change it also in some other packages.

**Comment by BKPepe** [comment]
For now, I think the package should be named python3-xmltodict as it is only Python3 package, but it's question if the folder should be named as well as python[3]-xmltodict or just xmltodict.

**Comment by jefferyto** [comment]
I'd like to mention some background info first:

* I remember adding the `python-` prefix for `PKG_NAME`s after 7a79f2db791751dd89bfa685831090c8b21897fb.
* Directory names are used as part of make target names (e.g. `make package/<dir-name>/compile`).

I'm not sure if these dependencies (`PKG_BUILD_DEPENDS` uses `PKG_NAME`, make target names use directory names) can be changed, but let's assume we are keeping them as-is.

I think we can be reasonably sure a Python library name is unique only among Python libraries (more specifically, ones available in PyPI). For example, there is only one xmltodict Python library, but there could also be a xmltodict Node library or a xmltodict PHP library.

So wherever a Python library name is used globally, I think we should scope it with a `python-` prefix. (I don't think having a version number in there is necessary, except when there is a need to distinguish between Python 2 and 3 versions.)

By this reasoning, because `PKG_NAME`s and directory names are used globally (in `PKG_BUILD_DEPENDS` and the make command line, respectively), these should have a `python-` prefix.

(Binary package names (`define Package/<name>`, `opkg install <name>`) are also global, and there is a need to distinguish between Python versions, so having `python-` and `python3-` make sense.)

I'd be interested to hear any other reasons for or against the prefix; I don't have any strong feelings either way. (Actually, I would prefer to lose the prefix for directory names, since all Python libraries are in the `python` directory, but I think the effort involved in changing the build system is greater than the strength of my annoyance :laughing:)

**Comment by commodo** [comment]
hmm; i thought I replied on this earlier;
oh well, need to rewrite

let's propose this:
* `lang/python/<pkg>`  - no python[3]- prefix
* `PKG_NAME:=python-<pkg>` - has `python-` prefix to avoid collisions
* `define Package/python3-<pkg>`  - same as before for Python3 / Python variants; evidently we don't allow python variants [unless really good reasons];

i'll check if it's `make package/<dir-name>/compile` or `make package/<PKG_NAME>/compile`;
if it is `make package/<dir-name>/compile`, then let's have ``lang/python/python-<pkg>` 

thoughts ?

strictly speaking for this PR, I'm fine for merging it as-is;
if there are issues, we can fix them later;

Thanks
Alex

**Comment by jefferyto** [comment]
@commodo I'm okay with that :+1: :joy:

**Comment by BKPepe** [comment]
The folder should be named exactly the same as it is in PKG_name to be more consistent with other packages in OpenWrt. However, it's a question for xmltodict. 

If I set PKG_NAME as python-xmltodict, I will need to override PKG_BUILD_DIR to look for folder xmltodict instead of python-xmltodict.

**Comment by commodo** [comment]
> If I set PKG_NAME as python-xmltodict, I will need to override PKG_BUILD_DIR to look for folder xmltodict instead of python-xmltodict.

Yep, I know. It is annoying.
I am working on a general change to impose PKG_BUILD_DIR.
See  https://github.com/commodo/openwrt/commit/81f43bb560e084ca9eda639da07d362ddf9c83c2
This needs a change in OpenWrt core that is a bit bigger than I thought initially, and requires more testing, but I will try to do it.

Python has had too many PKG_BUILD_DIR & PKG_UNPACK rules because of this.

**Comment by jefferyto** [comment]
I have a core PR open that is semi-related to this: openwrt/openwrt#1920 (seems no one is interested in it though :joy:)

I wanted to get rid of the PKG_BUILD_DIR / PKG_UNPACK rules for variants by standardizing what core packages (that have variants) were already doing on a per-package basis: adding an extra directory level, basically:

```
PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
```

so the default PKG_UNPACK rule still works (the base directory name is still `$(PKG_NAME)-$(PKG_VERSION)`)

I know this doesn't immediately help in this case (`PKG_NAME` is "python-x" but the extracted directory is "x"), but perhaps the two ideas can be combined somehow.

**Comment by commodo** [comment]
so, it seems that it is `make package/<pkg_dir>/compile`  and not `make package/<PKG_NAME>/compile`
i was never sure about this [ i know it sounds stupid ]

regarding this PR, it is fine as is; so, it can be merged ; sorry for the delay

maybe at a later point we could change the build-system [in a graceful manner] to do `make package/<PKG_NAME>/compile` 
until then, I'll try to not make much fuss about PKG_NAME and/or <pkg-dirname> to have a `python-` prefix [or not have it]

thanks :)
Alex

**Comment by commodo** [comment]
sorry for prolonging this longer than needed;

LGTM

thanks
Alex

**Comment by BKPepe** [comment]
No need to worry about it. I'm just waiting to buildpr to finish its job (of course with pass) as I don't want to merge anything while it is in progress.

**Pull request merged**